Propagate openmp to blas for sirius and spla

haampie opened the pull request:
This requires the clingo concretizer, with the old concretizer you run
into

```
sirius requires openblas variant threads=openmp, but spec asked for threads=none
```

because it's not smart enough to handle self-referential dependencies.
At this point the new concretizer can be enabled through

```
spack config edit config
```

by adding `concretizer: clingo`.

See also https://spack.readthedocs.io/en/latest/getting_started.html#optional-bootstrapping-clingo

haampie committed:
Propagate openmp to blas for sirius and spfft

This requires the clingo concretizer, with the old concretizer you run
into

```
sirius requires openblas variant threads=openmp, but spec asked for threads=none
```

because it's not smart enough to handle self-referential dependencies.
At this point the new concretizer can be enabled through

```
spack config edit config
```

by adding `concretizer: clingo`.

See also https://spack.readthedocs.io/en/latest/getting_started.html#optional-bootstrapping-clingo

haampie committed:
Remove lowerbounds on spfft/spla

AdhocMan commented on var/spack/repos/builtin/packages/spfft/package.py:
```diff
@@ -60,6 +60,12 @@ class Spfft(CMakePackage, CudaPackage):
     # FindHIP cmake script only works for < 4.1
     depends_on('hip@:4.0', when='@:1.0.1 +rocm')
 
+    # Propagate openmp to blas
```
SpFFT does not depend on BLAS at all, so I think these changes should be included for SPLA instead.

haampie replied:
Yes, that was intended for spla indeed.

AdhocMan commented on var/spack/repos/builtin/packages/sirius/package.py:
```diff
@@ -126,6 +123,16 @@ class Sirius(CMakePackage, CudaPackage):
+    depends_on('amdblis threads=openmp', when='+openmp ^amdblis')
+    depends_on('blis threads=openmp', when='+openmp ^blis')
+    depends_on('intel-mkl threads=openmp', when='+openmp ^intel-mkl')
+    depends_on('spfft+openmp', when='+openmp')
```
```suggestion
    depends_on('spfft+openmp', when='+openmp')
    depends_on('spla+openmp', when='+openmp')
```

toxa81 reviewed (dismissed):
This spack sorcery looks fine

haampie committed:
Move spfft blas constraints -> spla and improve

Make sirius depend on spfft+mpi

DRY the other depends_on statement to simplify things a bit.

haampie commented:
Can you have another look @AdhocMan?

haampie committed:
Fix style issue

AdhocMan commented on var/spack/repos/builtin/packages/sirius/package.py:
```diff
@@ -94,16 +94,15 @@ class Sirius(CMakePackage, CudaPackage):
-    depends_on('spfft@1.0.3:', when='@6.4.0:')
-    depends_on('spfft+cuda', when='@6.4.0:+cuda')
-    depends_on('spfft+rocm', when='@6.4.0:+rocm')
+    depends_on('spfft+mpi', when='@6.4.0:')
```
I just had a look again at the CMake config and the latest release requires 0.9.13. The next one will require 1.0.2, unless the requirement changes again.

AdhocMan commented on var/spack/repos/builtin/packages/sirius/package.py:
```diff
@@ -94,16 +94,15 @@ class Sirius(CMakePackage, CudaPackage):
-
-    depends_on('elpa+openmp', when='+elpa+openmp')
-    depends_on('elpa~openmp', when='+elpa~openmp')
+    depends_on('spla', when='@7.0.0:')
```
The same here, 1.1.0 is required for the latest release and possibly 1.4.0 in the future

AdhocMan reviewed (changes requested):
Looks mostly good to me now, but I'd suggest reverting the version requirement change for SpFFt and SPLA, since CMake config might fail for older versions.

haampie committed:
Fix lowerbounds for deps

haampie commented:
@AdhocMan I've updated the lowerbounds for tagged releases; I don't think it's useful to set them for master/develop, since those are moving targets.

AdhocMan reviewed (approved):
LGTM
